File: CPlus/SlidingWindowMaximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int> &nums, int k) {
        vector<int> ret;
        if(nums.empty())
            return ret;
        deque<int> q;
        for(int i=0; i<k; i++) {
            while(!q.empty() && nums[q.back()]<nums[i])
                q.pop_back();
            q.push_back(i);
            while(i - q.front() + 1 > k)
                q.pop_front();
        }
        
        ret.push_back(nums[q.front()]);
        //cout << "qsize=" << q.size() << endl;
        for(int i=k; i<nums.size(); i++) {
            //cout << "i=" << i << endl;
            while(!q.empty() && nums[q.back()]<nums[i])
                q.pop_back();
            q.push_back(i);
            while(i - q.front() + 1 > k) {
                //cout << "front=" << q.front() << endl;
                q.pop_front();
                //cout << "qsize=" << q.size() << endl;
            }
            ret.push_back(nums[q.front()]);
        }
        return ret;
    }
};

File: Python/slidingWindowMaximum.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        deque, r = collections.deque(), []
        for i, num in enumerate(nums):
            while deque and deque[0][1] <= i - k:
                deque.popleft()

            while deque and num > deque[-1][0]:
                deque.pop()

            deque.append([num, i])

            if i >= k - 1:
                r.append(deque[0][0])
        return r
